src/VisualStudio/Core/Def/Implementation/InheritanceMargin/InheritanceGlyphManager.cs

```diff
@@ -0,0 +1,210 @@
+        private readonly Canvas _glyphsContainer;
+
+        // We want to our glyphs to have the same background color as the glyphs in GlyphMargin.
+        private const string GlyphMarginName = "Indicator Margin";
+
+        // The same width and height as the margin of indicator margin.
+        private const double HeighAndWidthOfTheGlyph = 17;
+
+        public InheritanceGlyphManager(
+            IWpfTextView textView,
+            IThreadingContext threadingContext,
+            IStreamingFindUsagesPresenter streamingFindUsagesPresenter,
+            ClassificationTypeMap classificationTypeMap,
+            IClassificationFormatMap classificationFormatMap,
+            IUIThreadOperationExecutor operationExecutor,
+            IEditorFormatMap editorFormatMap,
+            IAsynchronousOperationListener listener,
+            Canvas canvas) : base(threadingContext, assertIsForeground: true)
+        {
+            _textView = textView;
+            _threadingContext = threadingContext;
+            _streamingFindUsagesPresenter = streamingFindUsagesPresenter;
+            _classificationTypeMap = classificationTypeMap;
+            _classificationFormatMap = classificationFormatMap;
+            _operationExecutor = operationExecutor;
+            _editorFormatMap = editorFormatMap;
+            _glyphsContainer = canvas;
+            _listener = listener;
+            _editorFormatMap.FormatMappingChanged += FormatMappingChanged;
+
+            _glyphToTaggedSpan = new Dictionary<InheritanceMarginGlyph, SnapshotSpan>();
+            UpdateBackgroundColor();
+        }
+
+        /// <summary>
+        /// Generate the glyph by the given <paramref name="tag"/>, and add it to the margin.
+        /// It should only be called by UI thread because UI elements are manipulated by this method.
+        /// </summary>
+        public void AddGlyph(InheritanceMarginTag tag, SnapshotSpan span)
+        {
+            AssertIsForeground();
+            var lines = _textView.TextViewLines;
+            if (lines.IntersectsBufferSpan(span) && GetStartingLine(lines, span) is IWpfTextViewLine line)
+            {
+                var glyph = CreateNewGlyph(tag);
+                glyph.Height = HeighAndWidthOfTheGlyph;
+                glyph.Width = HeighAndWidthOfTheGlyph;
+                SetTop(line, glyph);
+                _glyphToTaggedSpan[glyph] = span;
+                _glyphsContainer.Children.Add(glyph);
+            }
+        }
+
+        /// <summary>
+        /// Remove the glyphs covered by <paramref name="snapshotSpan"/>.
+        /// It should only be called by UI thread because UI elements are manipulated by this method.
+        /// </summary>
+        public void RemoveGlyph(SnapshotSpan snapshotSpan)
+        {
+            AssertIsForeground();
+            var marginsToRemove = _glyphToTaggedSpan
+                .Where(kvp => snapshotSpan.IntersectsWith(kvp.Value))
+                .ToImmutableArray();
```

Comment:
so, if this is cheap, this is fine.  however, if you have tons of these glyphs, and you want to do this efficiently, we have a really good datastructure called an IntervalTree which can be used to store tons of data, but query for a subset of them within a range.

Note that our core tagging infrastructure uses IntervalTrees for this purpose.  So if this system is part of our taggin system, it's a good chance you should be using htat.